### DISCUSSION TITLE
LG Magic Remote

### DISCUSSION THREAD

**Discussion by StratusFearMe21** [bluez/bluer#25] [General]
Hello, I am attempting to make a lightweight device driver for the LG Magic Remote. However, when I try to connect to it through `l2cap` it errors out with an `os error 111`. This is my code so far. It just tries to connect to the Magic Remote and it begins to receive packets.
```rust
use bluer::{
    l2cap::{SocketAddr, Stream, PSM_LE_DYN_START},
    Address,
};
use tokio::io::AsyncReadExt;

#[tokio::main]
async fn main() -> bluer::Result<()> {
    let addr: Address = "7C:02:BC:C5:14:3B".parse().unwrap();
    let socket_addr = SocketAddr::new(addr, bluer::AddressType::LePublic, PSM_LE_DYN_START + 5);
    println!("Connecting to magic remote");
    let mut stream = Stream::connect(socket_addr).await?;
    println!("Local address: {:?}", stream.as_ref().local_addr()?);
    println!("Remote address: {:?}", stream.peer_addr()?);
    println!("Recv MTU: {}", stream.as_ref().recv_mtu()?);
    println!("Security: {:?}", stream.as_ref().security()?);

    loop {
        let mut hello_buf = [0u8; 30];
        stream
            .read_exact(&mut hello_buf)
            .await
            .expect("read failed");
        println!("Received: {:?}", hello_buf);
    }
}
```
This is what it prints out
```
Connecting to magic remote
Local address: SocketAddr { addr: E0:D4:E8:38:4D:65, addr_type: LePublic, psm: 133, cid: 64 }
Remote address: SocketAddr { addr: 7C:02:BC:C5:14:3B, addr_type: LePublic, psm: 133, cid: 0 }
Recv MTU: 672
Security: Security { level: Medium, key_size: 0 }
thread 'main' panicked at 'read failed: Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }', src/main.rs:23:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Any ideas? Thanks!

**Comment by StratusFearMe21** [comment]
P.S there's no evidence of errors caused by app when I run it and then run `sudo systemctl status bluetooth`

**Comment by surban** [comment]
1. Are you sure that the LG Magic Remote is accepting L2CAP connections?
2. Are you sure that the port number is correct (`PSM_LE_DYN_START + 5` in your case)?

**Comment by StratusFearMe21** [comment]
Well what I do know, is that according to Wireshark, when I connect the magic remote to my laptop, the magic remote will constantly send packets to my computer that look like this
![bruh](https://user-images.githubusercontent.com/57533634/154261747-a91656c1-c2c1-4dd3-af91-192e246dac1d.png)
All I want to do is gain direct access to these packets.

**Comment by StratusFearMe21** [comment]
I suspect the protocol is as follows

- X Velocity > 64-bit floating point
- Y Velocity > 64-bit floating point
- Button currently pressed > 16 bit unsigned integer
- Null byte > 8 bit unsigned integer

**Comment by surban** [comment]
Please try running `gattcat discover [device-address]` and post its output here.

You can install gattcat using `cargo install bluer-tools`.
